### PR TITLE
[RFR] Refactoring to allow performance options

### LINKED ISSUE #1: Add a fast mode
Give https://github.com/Automattic/node-canvas#canvasrenderingcontext2dpatternquality and https://github.com/Automattic/node-canvas#canvasrenderingcontext2dfilter a try. If it's efficient, make it an option.

### PR DESCRIPTION
**Pull request by valeriangalliat**
Closes #1.

### CODE REVIEW

File: lib/cli.js [hunk 1]
```diff
@@ -2,9 +2,21 @@ const doc = `
...
 
 Options:
-  -h, --help             Bring help.
-  -v, --version          Show version.
-  -f, --format=<format>  Set output format ('png' or 'jpeg') [default: jpeg].
+  -h, --help                   Bring help.
+  -v, --version                Show version.
+  -f, --format=<format>        Set output format ('png' or 'jpeg')
+                               [default: jpeg].
+  -m, --mode=<mode>            Preset for the performance/quality options
```
Comment (KittyGiraudel) [inline lib/cli.js]: Shouldn’t it be `<quality>` here?

File: lib/cli.js [hunk 2]
```diff
@@ -2,9 +2,21 @@ const doc = `
...
+                               [default: good].
+  --pixel-interval=<interval>  Set pixel interval for average color
```
Comment (KittyGiraudel) [inline lib/cli.js:14]: Mention default?

File: lib/cli.js [hunk 3]
```diff
@@ -2,9 +2,21 @@ const doc = `
...
+                               computation (higher means faster but less
+                               accurate).
+  --pattern-quality=<quality>  Canvas pattern quality.
```
Comment (HugoGiraudel) [inline lib/cli.js]: Mention default?

File: lib/cli.js [hunk 4]
```diff
@@ -2,9 +2,21 @@ const doc = `
...
+  --filter=<quality>           Canvas filter quality (same values as pattern
```
Comment (HugoGiraudel) [inline lib/cli.js]: Mention default?

Comment (KittyGiraudel) [inline lib/cli.js]: Also, `filter-quality`.

File: lib/cli.js [hunk 5]
```diff
@@ -2,9 +2,21 @@ const doc = `
...
+                               quality).
+
+The options taking a quality argument can be one of 'fast', 'good', 'best',
```
Comment (KittyGiraudel) [inline lib/cli.js]: I’d go with `fast` and `good`. Everything else is just confusing.

File: lib/cli.js (hunk 1 above)
Comment (valeriangalliat) [inline lib/cli.js]: `--quality fast` ? Not sure.

File: lib/index.js [hunk 6]
```diff
@@ -2,42 +2,66 @@ import duplex from 'duplexer'
 import concat from 'concat-stream'
 import { PassThrough } from 'readable-stream'
 import Canvas from 'canvas'
-import brick from './brick'
 
-const getAverageColor = data => data
+const brick = __dirname + '/../images/brick.png'
```
Comment (KittyGiraudel) [inline lib/index.js]: Backticks. :)

File: lib/index.js [hunk 7]
```diff
@@ -2,42 +2,66 @@ import duplex from 'duplexer'
...
+const defaultOpts = { format: 'jpeg', mode: 'good', brick }
+const pixelIntervalPreset = { fast: 8, good: 5, best: 3, nearest: 2, bilinear: 1 }
+
+// Make a function to get the average color from given image data array
+// with given sample rate.
+const makeGetAverageColor = pixelInterval => data => data
+  .filter((_, index) => index % (pixelInterval * 4) < 4)
```
Comment (KittyGiraudel) [inline lib/index.js:13]: Why `_`?

File: lib/cli.js (hunk 2 above)
Comment (valeriangalliat) [inline lib/cli.js:14]: No default since preset takes precedence (same for other options).

File: lib/index.js [hunk 8]
```diff
@@ -2,42 +2,66 @@ import duplex from 'duplexer'
...
   .reduce((rgb, channel, index) => (rgb[index % 4] += channel, rgb), [0, 0, 0])
   .map(channel => Math.floor(channel / (data.length / 4)))
 
-const renderBrick = (ctx, x, y) => {
-  const color = getAverageColor(ctx.getImageData(x, y, brick.width, brick.height).data)
+// Make a function to render a brick at a point using given brick, average
+// color function and context.
+const makeRenderBrick = (brick, getAverageColor, ctx) => (x, y) => {
```
Comment (KittyGiraudel) [inline lib/index.js:19]: I don’t get this line. Why the double arrow?

File: lib/cli.js (hunk 4 above)
Comment (valeriangalliat) [inline lib/cli.js]: Why not. Was using the `canvas` option name.

File: lib/cli.js (hunk 5 above)
Comment (valeriangalliat) [inline lib/cli.js]: Well I'm just proxying the option to `canvas`, I don't want to loose configurability.

Though we could just document `fast` and `good` and leave a link.

File: lib/index.js (hunk 6 above)
Comment (valeriangalliat) [inline lib/index.js]: Yis!

File: lib/cli.js (hunk 1 above)
Comment (KittyGiraudel) [inline lib/cli.js]: Then we should mention allowed options.

File: lib/cli.js (hunk 2 above)
Comment (KittyGiraudel) [inline lib/cli.js:14]: Okay.

File: lib/index.js (hunk 7 above)
Comment (valeriangalliat) [inline lib/index.js:13]: Because unused. More or less a Rust convention.

File: lib/cli.js (hunk 4 above)
Comment (KittyGiraudel) [inline lib/cli.js]: Problem is consistency between `pattern-quality` and `filter`.

File: lib/cli.js (hunk 5 above)
Comment (KittyGiraudel) [inline lib/cli.js]: Agreed.

File: lib/index.js (hunk 8 above)
Comment (valeriangalliat) [inline lib/index.js:19]: So we can do `const renderBrick = makeRenderBrick(brick, getAverageColor, ctx)` and use the `renderBrick` function taking `x` and `y` only.

File: lib/index.js (hunk 7 above)
Comment (KittyGiraudel) [inline lib/index.js:13]: What is the JavaScript convention?

File: lib/index.js (hunk 8 above)
Comment (KittyGiraudel) [inline lib/index.js:19]: Okay, so it returns a function as I assumed. Good stuff.

File: lib/cli.js (hunk 1 above)
Comment (valeriangalliat) [inline lib/cli.js]: The same as `quality` below. Yeah... confusing. Will change.

File: lib/cli.js (hunk 4 above)
Comment (valeriangalliat) [inline lib/cli.js]: This is a problem with `canvas` options naming then. But we can fix it on our side, maybe.

File: lib/index.js (hunk 7 above)
Comment (valeriangalliat) [inline lib/index.js:13]: None in particular, but `_` is fairly common everywhere.

**Comment by KittyGiraudel** [comment]
Awesome PR. We just need to document it on the README maybe?

**Comment by valeriangalliat** [comment]
Sure.

File: lib/index.js (hunk 7 above)
Comment (KittyGiraudel) [inline lib/index.js:13]: Fair enough.

File: lib/cli.js [hunk 9]
```diff
@@ -2,9 +2,21 @@ const doc = `
...
+                               [default: good].
+  --pixel-interval=<interval>  Set pixel interval for average color
+                               computation (higher means faster but less
+                               accurate).
+  --pattern-quality=<quality>  Canvas pattern quality.
+  --filter-quality=<quality>   Canvas filter quality (same values as pattern
+                               quality).
+
+The options taking a quality argument can be one of 'fast', 'good', 'best',
+and 'nearest'.
```
Comment (KittyGiraudel) [inline lib/cli.js]: Fucking vim line breaks. :D

Comment (valeriangalliat) [inline lib/cli.js]: Not about Vim, it's about 80 chars.

File: lib/index.js [hunk 10]
```diff
@@ -2,42 +2,66 @@ import duplex from 'duplexer'
...
+const size = (image, brick) =>
+  [ image.width / brick.width, image.height / brick.height ].map(Math.floor)
+
+const render = (opts, src) => {
+  const image = img(src)
+  const brick = img(opts.brick)
+  const [ xCount, yCount ] = size(image, brick)
   const canvas = new Canvas(xCount * brick.width, yCount * brick.height)
   const ctx = canvas.getContext('2d')
+  const pixelInterval = opts.pixelInterval || pixelIntervalPreset[opts.quality]
```
Comment (KittyGiraudel) [inline lib/index.js:45]: I’d argue we don’t need this variable and can pass its value directly in the function call next line. Thought? Might be over 80 though.

Comment (valeriangalliat) [inline lib/index.js:45]: That's why!

**Pull request merged**